### DISCUSSION TITLE
[Berry] Callback inside class self being replaced

### DISCUSSION THREAD

**Discussion by Slyke** [arendst/Tasmota#15143] [Q&A] [answered]
Hello, I have the following code:
```
class MyClass
  var scr
  var btnClickCount
  var myButton1

  def init()
    lv.start()
    self.scr = lv.scr_act()
    self.myButton1 = lv.btn(self.scr)
    self.myButton1.add_event_cb(self.myCallback, lv.EVENT_CLICKED, 0)
    self.btnClickCount = 0
  end

  def myCallback(obj, event)
    self.btnClickCount = self.btnClickCount + 1
  end
end
```

When running this I get the following error:
```
BRY: Exception> 'attribute_error' - the 'lv_btn' object has no attribute 'btnClickCount'
stack traceback:
	<unknown source>: in function `myCallback`
	<unknown source>: in function `lvgl_event_dispatch`
	<unknown source>: in function `<lambda>`
```

I think the `self` reference is being replaced inside the callback, so I attempted this to fix it:
```
  self.myButton1.add_event_cb(def(obj, event) self.myCallback(self, obj, event) end, lv.EVENT_CLICKED, 0)
  ...
  def myCallback(parent, obj, event)
    parent.btnClickCount = parent.btnClickCount + 1
  end
```

But get this error in another part of the code (that I didn't include for brevity)
```
BRY: Exception> 'attribute_error' - 'nil' value has no method 'set_style_bg_color'
stack traceback:
	<unknown source>: in function `init`
	<unknown source>: in function `main`
	<unknown source>: in function `load`
	<unknown source>: in function `load`
	<unknown source>: in function `main`
```

I believe the second error is happening because the class' self was replaced.

**Comment by sfromis** [comment, marked as answer]
You cannot use `self` in a callback function, as it does not get access the the class instance. What you can do is to use a closure where you do the add. That closure will get access to `self`, and thus can call your instance function the usual way, not including `self` as a parameter to the function, as this is done automatically behind the curtain. Abbreviated notation for closure:
```
  self.myButton1.add_event_cb(/obj, event -> self.myCallback(obj, event), lv.EVENT_CLICKED, 0)
  ...
  def myCallback(obj, event)
    self.btnClickCount = self.btnClickCount + 1
  end
```

If you wanted to continue the avenue passing a parent parameter, the callback function would have to be static, but still a closure to have access to `self` for doing the call.

**Reply by Slyke** [reply]
Thank  you! That worked.
